plugins/module_utils/azure_rm_common.py

```diff
@@ -1391,10 +1409,13 @@ def _get_msi_credentials(self, subscription_id_param=None, **kwargs):
             'subscription_id': subscription_id
         }
 
-    def _get_azure_cli_credentials(self, resource=None):
+    def _get_azure_cli_credentials(self, subscription_id=None, resource=None):
         if self.is_ad_resource:
             resource = 'https://graph.windows.net/'
-        credentials, subscription_id = get_azure_cli_credentials(resource)
+        subscription_id = subscription_id or self._get_env('subscription_id')
```

Comment:
@UnwashedMeme Why do you catch the subscription_id of the environment variable here? auth_source='cli' uses the credentials of 'az loggin', so the environment variable subscription should not be used in this function, please update. Thank you!

Reply:
> why do you catch the subscription_id of the environment variable here?

Because that is a valid way of providing data to these modules and this PR reduces the number of cases in which a user provided subscription_id is unexpectedly ignored.

More broadly *"The authentication mechanism (service principal, msi, cli) and parameter passing convention (module parameters, environment variables, profile file) should be as loosely coupled as possible."* 

I believe you have been looking at this as "There currently several distinct parameter sets that are entirely separate depending upon which `auth_source` is specified."  I don't believe that is true in existing code except by accident and it certainly isn't stated in the module documentation. 

The only places subscription_id should be required is when we don't have a mechanism for looking it up associated with the authentication.   The [service principal authentication code](https://github.com/ansible-collections/azure/blob/dev/plugins/module_utils/azure_rm_common.py#L1315) does not use the subscription_id. But since we need to have a subscription_id [when creating the mgmt client](https://github.com/ansible-collections/azure/blob/dev/plugins/module_utils/azure_rm_common.py#L851) it is required.  When not set explicitly, some mechanisms like MSI and CLI can provide a default. _If the module user provides a subscription id through any mechanism it should be respected._

Reply:
Yes, but if you assign the subscription_id of the environment variable here, then auth_source= cli will not use the credentials of 'az logging', which is contrary to the designed auth_source parameter.

Reply:
So I'm think here, if you set ’auth_source= cli‘， then you're going to use the 'az login' credentials, and if you can't catch the credentials or if the credentials are wrong, you're going to throw an exception,' az Login 'credentials have a problem. Thank you very much!

Reply:
>  if you assign the subscription_id of the environment variable here, then auth_source= cli will not use the credentials of 'az logging'

This is not a problem because `subscription_id` is not part of the credentials from `az login`.  Please recall that the credentials from az login can grant access to many subscriptions. The subscription_id module parameter or environment variable just allows the user to specify which one this module should actually target.

As with _any_ auth_source, if the user specifies a subscription_id that the credentials aren't authorized for it won't work.

Reply:
Yes, I'll also update my description, the subscription_id here should get the subscription_id of the playbook parameter, not the one in the environment variable. Thank you very much!